File: Solutions/3. Longest Substring Without Repeating Characters/Python/Longest_Substring_without_Repeating_Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        s_length = len(s)

        #init length upperbound, starting from each char
        visit = {}
        len_ub = [float('inf')] * s_length
        for i in range(1, s_length + 1):
            if s[-i] in visit: #c is visited
                len_ub[-i] = i - visit[s[-i]]

            visit[s[-i]] = i

        #calculate maxLen
        maxLen = 0
        for i in range(s_length):
            if s_length - i < maxLen:
                break

            stepCount  = 0
            remainStep = float('inf')
            for j in range(i, s_length):
                if remainStep == 0:
                    break
                
                #update remainStep
                remainStep = min(remainStep - 1, len_ub[j] - 1)
                stepCount  = stepCount  + 1

            maxLen = max(maxLen, stepCount)
    
        return maxLen

S = Solution()

#input
s = "abcabcbb"

print(S.lengthOfLongestSubstring(s))

File: Solutions/3. Longest Substring Without Repeating Characters/CPP/Longest_Substring_without_Repeating_Characters.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.empty()) {
            return 0;
        }

        int result = 1;

        int s_length = s.length();
        vector<int> lengthCounter(s_length, 1);

        // lengthCounter
        for (int i = 1; i < s_length; ++i) {
            int intervalWidth = i;
            bool isDone = true;
            for (int j = 0; j < s_length - intervalWidth; ++j) {
                if ((lengthCounter[j] == intervalWidth) && (lengthCounter[j + 1] == intervalWidth) && (s[j] != s[j + intervalWidth])) {
                    lengthCounter[j] = lengthCounter[j] + 1;
                    result = lengthCounter[j];
                    isDone = false;
                }
            }
            if (isDone) {  // no merge happened
                break;
            }
        }

        return result;
    }
};

int main() {
    Solution S;

    // input
    string s = "aaabcdeababb";

    cout << S.lengthOfLongestSubstring(s) << endl;

    return 0;
}
